### PR TITLE
fix bug in plot rgb when user forgets to deal with no data values + bytescale fix + test cleanup

### PR DESCRIPTION
**Pull request by lwasser**
Please see https://github.com/earthlab/earthpy/issues/282 

When a user provides and RGB image to plot_rgb but leaves the no data values in, the result is an error with bytescale if the no data val is small (e.g. -9999). Here, we check to ensure the max-min is not less than 0 BEFORE bytescale fails on them. We will implement this in conjunction with an upgrade to `stack_bands` to allow for no data values in hopes users won't forget to handle na values when plotting. 

UPDATE: through this pr i will

1. fix a bug in bytescale to ensure range is calculating properly via floating point vals.
2. break up bytescale tests into a new file AND also individual tests - they are currently all in one ginormous function
3. add some notes to the docs for plot_rgb to ensure users clean our no data values.

### CODE REVIEW

**Comment by lwasser** [comment]
weird things
1. this test
```
def test_1band(rgb_image):
    """Test to ensure the input image has 3 bands to support rgb plotting.
    If fewer than 3 bands are provided, fail gracefully."""
    a_rgb_image, _ = rgb_image

    with pytest.raises(ValueError, match="Input needs to be 3 dimensions"):
        plot_rgb(a_rgb_image[1])
    plt.close()
``` 
tests for a value error yet [code cov](https://codecov.io/gh/earthlab/earthpy/pull/287/src/earthpy/plot.py?before=earthpy/plot.py#L299) is noticing that the message part of the error is NOT tested. not sure why. the function name is unique.

@mbjoseph in theory this pr is good to go. code cov is not happy and i'm not sure why. the above is one line that it thinks is not being hit but is being hit (i think?). anyway please have a look and let me know what you think!!

File: earthpy/plot.py [hunk 1]
```diff
@@ -327,7 +328,8 @@ def plot_rgb(
     ax.imshow(rgb_bands, extent=extent)
     ax.set_title(title)
     ax.set(xticks=[], yticks=[])
-    return fig, ax
```
Comment (mbjoseph) [inline earthpy/plot.py]: This seems like a substantial change in behavior - is this related to #282? 

If users are trying to use this function to make multipanel plots or use the `fig` object, will this change prevent them from doing so?

File: earthpy/tests/test_bytescale.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import earthpy.spatial as es
+
+
+@pytest.fixture
+def byte_arr():
+    return np.random.randint(300, size=(10, 10))
+
+
+def test_high_val_range():
+    """A 16 bit int arr with values at the end of the range, should be scaled properly.
```
Comment (mbjoseph) [inline earthpy/tests/test_bytescale.py]: Each line in this docstring goes over the 79 character limit

File: earthpy/tests/test_bytescale.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    Note that this test explicitly hits the user case of someone providing an array of a
+    dtype at the end of it's range of values. Bytescale should not fail in this case.
+    """
+
```
Comment (mbjoseph) [inline earthpy/tests/test_bytescale.py]: Good place for a comment to explain where these magic numbers come from
```suggestion
    # The range of int16 values in python is -32768 to 32767
```

File: earthpy/tests/test_bytescale.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    # High value is less than low value
+    with pytest.raises(
+        ValueError, match="`high` should be greater than or equal to `low`."
+    ):
+        es.bytescale(byte_arr, high=100, low=150)
+
+
+def test_low_high_vals_work(byte_arr):
+    """A proper high and low value argument should return an arr with thos vals as the min and max."""
```
Comment (mbjoseph) [inline earthpy/tests/test_bytescale.py]: This seems like it could be condensed

```suggestion
    """ The high and low args set the output min and max. """
```

File: earthpy/tests/test_bytescale.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    ):
+        es.bytescale(byte_arr, cmin=100, cmax=50)
+
+
+def test_cmax_equals_cmin(byte_arr):
+    """Fail gracefully when the cmax is smaller than the cmin."""
+
+    with pytest.raises(
+        ValueError,
+        match="`cmax` and `cmin` should not be the same value. Please specify `cmax` > `cmin`",
```
Comment (mbjoseph) [inline earthpy/tests/test_bytescale.py]: This line is longer than it needs to be - `match` uses regex so we can do:

```suggestion
        match="`cmax` and `cmin` should not be the same value",
```

File: earthpy/tests/test_bytescale.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    ):
+        es.bytescale(byte_arr, cmin=100, cmax=100)
+
+
+def test_cmax_cmin_work(byte_arr):
+    """"Cmax and min values should yield and arr with the same min and max values."""
```
Comment (mbjoseph) [inline earthpy/tests/test_bytescale.py]: Can also be condensed and clarified:

```suggestion
    """" cmin and cmax should not change the output min and max. """
```

File: earthpy/tests/test_plot_rgb.py [hunk 7]
```diff
@@ -26,14 +26,25 @@ def rgb_image():
     return rgb, ext
 
 
+def test_no_data_val(rgb_image):
+    """An array with a large no data value should not print nicely, but it should work."""
```
Comment (mbjoseph) [inline earthpy/tests/test_plot_rgb.py]: This docstring doesn't really describe what's being tested (e.g., we aren't testing that the array is not printing nicely).

Can it be clarified and shortened to fit with our line length limits? e.g., 

```suggestion
    """ Arrays containing -9999 make plots when stretched. """
```

File: earthpy/plot.py (hunk 1 above)
Comment (lwasser) [reply earthpy/plot.py]: Nope. 
1. you can always access the figure object using ax.figure if need be.
BUT if a user wants to implement multi-panel, they need to setup the plot themselves and then use the `ax=` parameter ... 

this test
```
def test_ax_provided(rgb_image):
```
tests that a user can in fact provide a custom axis and things still work!

Comment (mbjoseph) [reply earthpy/plot.py]: Sweet!

**Comment by lwasser** [comment]
ok @mbjoseph travis looks happy - the github web hook doesn't seem to be updating but when i go to the travis build it is passing. I think i've covered all of your feedback here but let me know. I wasn't sure what you meant about `plt.show()` as i am calling it in the function... but if you help me better understand what you are looking for i can likely account for it!!

**Comment by mbjoseph** [comment]
Looks good @lwasser!

**Comment by lwasser** [comment]
yahoo! thanks @mbjoseph !!!
